### PR TITLE
Update engine api with required arguments

### PR DESCRIPTION
**Pull request by vdemeester**
This PR is related to docker/engine-ap#162 and is it's inclusion in `docker/docker`. _It's still a little bit a work in progress_ but I wanted to have both PR opened at the same time to make sure it's ok for everybody and to measure the impact of the change on the code here :angel:.

The docker/engine-api#162 (docker/engine-api#137 for the related issue) changes some `client` signature to add the required arguments in the signature (to make it easier to use by telling the consumer what is required and what is not) 🐯.

It also tries to remove the `image`/`tag` distinction from the client side for now, and replace it with the `reference` instead. 
This has some impact on the content trust code (`api/client/trust.go`). 

This is also a very small step to remove the `reference` package (or at least reduce the usage) from this repository (and use `docker/distribution/reference` in the future, once docker/distribution#1277 is merged) 🐳.

/cc @calavera @stevvooe @MHBauer @dmcgowan @aaronlehmann @tonistiigi @LK4D4 

It's a little bit a chicken-and-egg problem, as depending on the review here, the code in the API will have to be updated. Anyway, we have to wait for docker/engine-api#162 to get merged first but we will merge it once it's green, reviewed and OK here :angel: .

_Let's hope it's green first :sweat_smile:_. 

🐸

### CODE REVIEW

File: api/client/import.go [hunk 1]
```diff
@@ -31,26 +30,20 @@ func (cli *DockerCli) CmdImport(args ...string) error {
...
+		in      io.Reader
+		tag     string
+		src     = cmd.Arg(0)
+		srcName = src
+		ref     = cmd.Arg(1)
+		changes = flChanges.GetAll()
 	)
 
 	if cmd.NArg() == 3 {
+		// FIXME(vdemeester) Which version has this been deprecated ? should we remove it ?
```
Comment (vdemeester) [inline api/client/import.go]: This message is there since at least `1.8`, so I think it could be removed. If yes, then the `tag` variable is no more needed and the `Tag` attribute in `ImageImportOptions` neither.

**Comment by thaJeztah** [comment]
Looks like CI is failing:

```
16:15:28 + go test -check.f DockerSuite -check.v -check.timeout=90m -test.timeout=360m github.com/docker/docker/integration-cli
16:15:57 # github.com/docker/docker/integration-cli
16:15:57 ./docker_api_network_test.go:32: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:47: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:51: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:103: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:129: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:173: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:186: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:199: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:212: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:214: unknown "github.com/docker/engine-api/types".NetworkCreate field 'Name' in struct literal
16:15:57 ./docker_api_network_test.go:214: too many errors
```

File: api/client/push.go [hunk 2]
```diff
@@ -30,14 +28,6 @@ func (cli *DockerCli) CmdPush(args ...string) error {
 		return err
 	}
 
-	var tag string
-	switch x := ref.(type) {
-	case reference.Canonical:
-		return errors.New("cannot push a digest reference")
```
Comment (aaronlehmann) [inline api/client/push.go]: Why is this removed?

Comment (vdemeester) [inline api/client/push.go]: oh you are right, something is missing in the `ImagePush` in remote API (same as `ImageTag` in fact)

**Comment by aaronlehmann** [comment]
I like the direction this is going. My main question is about references being passed as strings into engine-api functions. Would it be better to use the actual `distribution/reference` types? I see cases where we're taking a parsed reference and calling `String` before passing it to an API function, and then the API function parses it again. I also see arguments like `repoInfo.Name()+"@"+r.digest.String()`.

I think there are advantages to passing in actual references to the API functions. It prevents you from making mistakes like passing in a tag where you're really supposed to pass a fully-qualified reference.

Are there reasons not to expose these types at the API level? Or do we plan to do that, with this PR as an intermediate step on the way there?

@stevvooe WDYT?

**Comment by stevvooe** [comment]
@aaronlehmann I'm not sure that the `distribution/reference` types are ready for prime time. We are still seeing issues with their use in `docker/docker`. While they provide correctness, they are slightly clunky for common API use. I'd rather people just take a string user input and shove it into this function, where we can validate and parse it correctly.

> I think there are advantages to passing in actual references to the API functions. It prevents you from making mistakes like passing in a tag where you're really supposed to pass a fully-qualified reference.

> Are there reasons not to expose these types at the API level? Or do we plan to do that, with this PR as an intermediate step on the way there?

I agree 100% with the sentiment here but we may want to consider this PR to be an intermediate step.

Ideally, these all become completely server-side, avoiding the need to parse and validate them client-side.

**Comment by vdemeester** [comment]
> Are there reasons not to expose these types at the API level? Or do we plan to do that, with this PR as an intermediate step on the way there?

I agree with @stevvooe, "*Ideally, these all become completely server-side, avoiding the need to parse and validate them client-side.*". It make it simpler for a user of the `engine-api` to use it (like in libcompose) and it's *kinda* consistent with the cli (in the way that you specify the image reference as a "string" on its own on the cli, should be the same using the api).

And yes, I see this PR as a first step on moving towards having the reference parsing handling server-side (but it's a little more work and it would make the PR bigger).

**Comment by vdemeester** [comment]
Updated the code (here and in the engine-api PR) and fixed the build :smile_cat:

**Comment by calavera** [comment]
@vdemeester the change in engine-api was merged this morning. Feel free to update this PR. Moving to code review.

**Comment by vdemeester** [comment]
Win2Lin is green :
```
11:15:29 23cd6eeee6e6
11:15:39 INFO: Cleanup complete
11:15:39 Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
11:15:39 Finished: SUCCESS
```

Userns is not :sweat: 

```
11:11:50 ----------------------------------------------------------------------
11:11:50 FAIL: docker_cli_run_unix_test.go:808: DockerSuite.TestRunSeccompProfileDenyUnshareUserns
11:11:50 
11:11:50 docker_cli_run_unix_test.go:839:
11:11:50     c.Fatalf("expected unshare userns with seccomp profile denied to fail, got %s", out)
11:11:50 ... Error: expected unshare userns with seccomp profile denied to fail, got /go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/../binary/docker: Error response from daemon: Untar re-exec error: exit status 1: output: no such file or directory.
11:11:50 See '/go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/../binary/docker run --help'.
11:11:50 
11:11:50 
11:11:50 
11:11:50 ----------------------------------------------------------------------
```

**Comment by calavera** [comment]
LGTM

File: api/client/push.go (hunk 2 above)
Comment (stevvooe) [inline api/client/push.go]: Is this resolved?

File: api/client/trust.go [hunk 3]
```diff
@@ -377,7 +373,7 @@ func (cli *DockerCli) trustedPull(repoInfo *registry.RepositoryInfo, ref registr
 		}
 		fmt.Fprintf(cli.out, "Pull (%d of %d): %s%s@%s\n", i+1, len(refs), repoInfo.Name(), displayTag, r.digest)
 
-		if err := cli.imagePullPrivileged(authConfig, repoInfo.Name(), r.digest.String(), requestPrivilege); err != nil {
+		if err := cli.imagePullPrivileged(authConfig, repoInfo.Name()+"@"+r.digest.String(), requestPrivilege); err != nil {
```
Comment (stevvooe) [inline api/client/trust.go]: Do we not have a way to assemble canonical references? https://godoc.org/github.com/docker/distribution/reference#WithDigest

May have to use the docker/docker/reference version.

File: vendor/src/github.com/docker/engine-api/client/privileged.go [hunk 4]
```diff
@@ -1,9 +1 @@
 package client
```
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/client/privileged.go]: Was this entire file not deleted? I'll file a PR...

Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/client/privileged.go]: Filed docker/engine-api#201. Don't hold this PR up on merging that.

File: api/client/push.go (hunk 2 above)
Comment (dmcgowan) [inline api/client/push.go]: Looks like it was resolved https://github.com/docker/engine-api/pull/162/files#diff-af7c2f642033c3058dc4c26810d875a4R27

File: api/client/trust.go (hunk 3 above)
Comment (vdemeester) [inline api/client/trust.go]: Hum you're right

**Comment by vdemeester** [comment]
@stevvooe updated (using `WithDigest` from `docker/docker/reference` for now – as it's already used elsewhere in this file)

**Comment by stevvooe** [comment]
LGTM
